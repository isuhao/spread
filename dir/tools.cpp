#include "tools.hpp"

using namespace Spread;

typedef Hash::DirMap DirMap;

void Spread::Dir::add(DirMap &dir, const DirMap &d, const std::string &prefix)
{
  for(DirMap::const_iterator it = d.begin(); it != d.end(); it++)
    dir[prefix+it->first] = it->second;
}
