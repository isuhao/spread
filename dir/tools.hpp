#ifndef __SPREAD_DIR_TOOLS_HPP_
#define __SPREAD_DIR_TOOLS_HPP_

#include "hash/hash.hpp"

namespace Spread
{
  namespace Dir
  {
    /* Meld dir2 into dir. An optional prefix can be added to each
       filename.
     */
    extern void add(Hash::DirMap &dir, const Hash::DirMap &dir2,
                    const std::string &prefix = "");
  }
}
#endif
