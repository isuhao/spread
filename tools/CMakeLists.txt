cmake_minimum_required(VERSION 2.6)

include_directories("../")
include_directories("../../libs/")

set(HASH ../hash/hash.cpp ../libs/sha2/sha2.c ../misc/comp85.cpp)

add_executable(spreadsum spreadsum.cpp ${HASH})
