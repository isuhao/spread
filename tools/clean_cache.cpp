#include <iostream>
#include <cache/index.hpp>

using namespace std;
using namespace Cache;
using namespace Spread;

int main(int argc, char **argv)
{
  if(argc != 2)
    {
      cout << "Syntax: clean_cache <cache-file>\n";
      return 1;
    }

  string file = argv[1];
  cout << "Loading " << file << "\n";
  CacheIndex cache(argv[1]);
  cout << "Cleaning...\n";
  cache.verify();
  cout << "Done!\n";

  return 0;
}
