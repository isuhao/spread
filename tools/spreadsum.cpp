#include <hash/hash.hpp>

#include <fstream>
#include <vector>
#include <iostream>
#include <stdexcept>

using namespace std;
using namespace Spread;

Hash hashStream(istream &inf)
{
  vector<char> buf(1024);
  Hash res;

  while(!inf.eof())
    {
      // Read from input file
      inf.read(&buf[0], buf.size());
      if(inf.bad())
        return Hash();

      int count = inf.gcount();

      // Hash the read data
      res.update(&buf[0], count);
    }

  // We're done, return the hash
  return res.finish();
}

Hash hashFile(const std::string &src)
{
  // Input file
  ifstream inf(src.c_str(), ios::binary);
  if(!inf) return Hash();

  return hashStream(inf);
}

int main(int argc, char**argv)
{
  bool json = false;
  bool cleanup = false;

  /*
    0 = hex
    1 = base64
   */
  int format = 1;

  vector<string> files;

  // Search through parameters
  for(int i=1; i<argc; i++)
    {
      string par = argv[i];

      if(par == "-j")
        json = true;
      else if(par == "-c")
        cleanup = true;
      else if(par == "-b")
        format = 1;
      else if(par == "-x")
        format = 0;
      else if(par == "--help")
        {
          cout << "spreadsum [options] <files>\n\n";
          cout << "  -j    Output in JSON format\n";
          cout << "  -c    Remove leading ./ from filenames\n";
          cout << "  -x    Output in hex format\n";
          cout << "  -b    Output in base64 format (default)\n";
          return 0;
        }
      else
        files.push_back(par);
    }

  if(files.size() == 0)
    files.push_back("-");

  if(json)
    cout << "{\n";

  for(int i=0; i<files.size(); i++)
    {
      string file = files[i];

      if(cleanup && file.substr(0,2) == "./")
        file = file.substr(2);

      Hash h;
      if(file == "-")
        h = hashStream(cin);
      else
        h = hashFile(file);

      string hstring;

      if(format == 0)
        hstring = h.toHex();
      else if(format == 1)
        hstring = h.toString();

      if(json)
        {
          cout << "    \"" << file << "\":\"" << hstring << "\"";
          if(i+1 != files.size())
            cout << ",";
          cout << "\n";
        }
      else
        {
          // Pad compact values to look nice
          for(int i=hstring.size(); i<50; i++)
            hstring += " ";
          cout << hstring << "  " << file << endl;
        }
    }

  if(json)
    cout << "}\n";

  return 0;
}
