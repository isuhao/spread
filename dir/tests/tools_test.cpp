#include "tools.hpp"
#include <iostream>

using namespace std;
using namespace Spread;

void print(const Hash::DirMap &dir)
{
  Hash::DirMap::const_iterator it;
  for(it = dir.begin(); it != dir.end(); it++)
    {
      const std::string &file = it->first;
      const Hash &hash = it->second;
      assert(file != "");
      cout << hash << " " << file << endl;
    }
  cout << endl;
}

int main()
{
  Hash::DirMap dir, sum;

  dir["abc"] = Hash("ABC");
  dir["def"] = Hash("DEF");

  Dir::add(sum, dir);
  print(sum);

  Dir::add(sum, dir, "prefix_");
  print(sum);

  dir["abc"] = Hash("GHIIIII");
  Dir::add(sum, dir);
  print(sum);

  return 0;
}
