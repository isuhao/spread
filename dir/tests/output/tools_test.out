ABC abc
DEF def

ABC abc
DEF def
ABC prefix_abc
DEF prefix_def

GHIIIII abc
DEF def
ABC prefix_abc
DEF prefix_def

